charlib/liberty/liberty.py

```
@@ -16,11 +18,28 @@ def name(self, name):
 
 
 class Group(Statement):
+
+    class SubGroupDict(UserDict):
+        """Custom dictionary with auto-updating keys. Used for keeping track of subgroups."""
+
+        def __setitem__(self, key, value):
+            self._update_key(key, value)
+            value._bind_to_parent_group(self)
+
+        def _update_key(self, key, value):
```

Ruff (0.15.20)

[warning] Missing return type annotation for private function `_update_key`

Add return type annotation: `None`

(ANN202)